Convert kubelet metrics(running_pod_count and running_container_count) from non-standard prometheus collectors to standard gauges

irajdeep opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind design


**What this PR does / why we need it**:
This PR attempts to fix [this issue](https://github.com/kubernetes/kubernetes/issues/79286). It converts kubelet metrics (e.g running_pod_count and running_container_count)from non-standard prometheus collecter to standard gauge metrics

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
https://github.com/kubernetes/kubernetes/issues/79286

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->

```release-note
Add `container_state` label to `running_container_count` kubelet metrics, to get count of containers based on their state(running/exited/created/unknown)
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

irajdeep committed:
Update running_pod_count and running_container_count metric

As already mentioned in this issue https://github.com/kubernetes/kubernetes/issues/79286, some metrics like
"running_pod_count" and "running_container_count" uses non-standard prometheus metrics, this change converts them to be
standard prometheus gauges

Minor refactor in kubelet/pleg/generic.go and added some test for ruuning container and running pod metrics

Fixed issues related to github CI pipeline failure

* Updated bazel for new deps
* Add comment for exported metrics variables,RuuningContainerCount and RunningPodCount
* Specify keys explicitly in Guage metric instantation

Fix go lint errors

Replace "+=1" with "++", as reported by go lint

Set container state as a label for the metrics "running_container_count"

As per the metrics name "running_container_count" it should "ideally" be showing
the number of containers in "running" state , but it was showing all the container count, irrespective of the state it is in.
This commit adds a new label "container_running_state" to the metrics "running_container_count", which doesn't change the base metrics but adds the
option to query the metrics with "container_state" such as "running"/"unknown/...

remove unused methods reported by staticcheck

Remove variables while instantiating gauge(vec) which are default set to nil

Convert kubelet metrics(running_pod_count and running_container_count) to standard gauges and added label to running_container_count metrics.

Currently kubelet metrics(running_pod_count and running_container_count) use non-standard prometheus collectors , this change
converts them to standard prometheus gauges. Also this adds a new label(container_state) to running_container_count which does a breakdown of
containers tracked by kubelet based on the containers' state(running/unknown/created/exited).

Set statbility explicitly for running_pod_count and running_container_count and reformat test

register metrics explicitly in test , so that they don't become no-op

irajdeep commented:
/assign logicalhan

logicalhan commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -431,6 +433,24 @@ func getContainerState(pod *kubecontainer.Pod, cid *kubecontainer.ContainerID) p
+		containers := pod.Containers
+		for _, container := range containers {
+			// check if the state of the container is "running"
+			if container.State == kubecontainer.ContainerStateRunning {
```
This logic is a bit different than the existing logic for the metric (which does a simpler `len(p.Containers)`). @dashpole, for more input.

irajdeep replied:
Right, I did this change because it aligns more with the logic of metric name, i.e counting the containers in the running state.

logicalhan replied:
Yeah, I agree it makes more sense; regardless, it is a change in the semantics of the existing metric so we should check with the node team to ensure it isn't likely to break anything.

dashpole replied:
I don't think we can change it.  Its hard to reason about what can break.  Could we add the container state as a label instead?  That wouldn't change queries over the base metric, and allow doing the filtering at a later point.

irajdeep replied:
That makes total sense. 
I have now added the container state as a label `container_state` in the `running_container_count` metrics.

logicalhan reviewed (commented):
lgtm but @dashpole for more input on the change in the metric.

logicalhan commented:
/assign @dashpole

logicalhan commented:
/sig instrumentation

logicalhan commented:
/ok-to-test

irajdeep commented:
/test pull-kubernetes-bazel-test

jan25 commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -431,6 +433,24 @@ func getContainerState(pod *kubecontainer.Pod, cid *kubecontainer.ContainerID) p
+	// get number of running containers given a Pod
+	numberRunningContainers := 0
+	for _, pod := range pods {
+		containers := pod.Containers
```
nit: variable not necessary

jan25 commented:
thanks! Please fix CI

irajdeep commented:
Okay, so tests seem to be passing now.
@dashpole could you have a look and provide some inputs.

Thanks

logicalhan reviewed (commented):
Some minor nits, but it looks mostly good to me, would you mind squashing your changes?

brancz commented:
/lgtm

irajdeep commented:
> Some minor nits, but it looks mostly good to me, would you mind squashing your changes?

Done. Fixed the minor issues and squashed the commits.

irajdeep commented:
Good point. Updated PR description with release note

feiskyer reviewed (commented):
/lgtm

logicalhan commented:
You will need to rebase as the kubelet migration changes went in first. That means you'll have to change your registration calls to use the migration registry ones.

irajdeep commented:
/test pull-kubernetes-e2e-gce

irajdeep commented:
> You will need to rebase as the kubelet migration changes went in first. That means you'll have to change your registration calls to use the migration registry ones.

Done!

derekwaynecarr commented:
/hold cancel
/approve

derekwaynecarr commented:
/milestone v1.16

logicalhan commented:
/lgtm

tedyu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -431,6 +433,24 @@ func getContainerState(pod *kubecontainer.Pod, cid *kubecontainer.ContainerID) p
+	containerStateCount := make(map[string]int)
+
+	for _, pod := range pods {
+		containers := pod.Containers
```
Should InitContainers be included ?

irajdeep replied:
Makes sense introducing it in a new GaugeVec(`running_initcontainer_count` something like this maybe)
@logicalhan what do you think?

tedyu replied:
I found that container.Pod doesn't expose initcontainers.
Edited comment above for Sandboxes.

If this makes sense, I can send out a PR.

dashpole replied:
We could add this, although i'm not 100% sure what the value of it would be.  It should be exactly equal to running pod count except in rare scenarios when the kubelet is recreating a sandbox.  But i'm not entirely sure why knowing that a sandbox is being recreated is useful... What did you intend to use it for?

tedyu replied:
Assuming the sandbox count settles down after the recreation, probably we don't need the extra gauge.
